controllers/release/release_adapter.go

```diff
@@ -228,24 +229,39 @@ func (a *Adapter) EnsureSnapshotEnvironmentBindingIsCreated() (results.Operation
 // EnsureTargetContextIsSet is an operation that will ensure that the targetContext of the adapter is not nil. It will
 // set it to a context from the workspace in the ReleasePlan if one is defined or to its own context if not.
 func (a *Adapter) EnsureTargetContextIsSet() (results.OperationResult, error) {
-	if a.targetContext == nil {
+	// If the target namespace/workspace is not set at the release status we should get the target from the ReleasePlan
+	if a.release.Status.Target == (kcp.NamespaceReference{}) {
+		patch := client.MergeFrom(a.release.DeepCopy())
 		releasePlan, err := a.getReleasePlan()
 		if err != nil {
-			patch := client.MergeFrom(a.release.DeepCopy())
 			a.release.MarkInvalid(v1alpha1.ReleaseReasonReleasePlanValidationError, err.Error())
 			return results.RequeueOnErrorOrStop(a.client.Status().Patch(a.context, a.release, patch))
 		}
 
-		if releasePlan.Spec.Target.Workspace == "" {
-			a.logger.Info("Running in a traditional kubernetes environment")
-			a.targetContext = a.context
-		} else {
-			a.logger.Info("Running in a KCP environment")
-			a.targetContext = logicalcluster.WithCluster(a.context, logicalcluster.New(releasePlan.Spec.Target.Workspace))
+		a.release.Status.Target = releasePlan.Spec.Target
+		err = a.client.Status().Patch(a.context, a.release, patch)
+		if err != nil {
+			return results.RequeueWithError(err)
 		}
 
+	}
+
+	if a.release.Status.Target.Workspace != "" {
+		// If this is a non-KCP cluster and the target workspace is set we should stop the release
+		// otherwise it fails at the release_controller because it can not switch to a non existent
+		// context
+		if _, ok := logicalcluster.ClusterFromContext(a.context); !ok {
```

Comment:
This block should be moved to the first condition where we set the target in the Release status. We don't want to check this (nor log that we are running in a KCP environment) in every reconciliation.